mantidproject/mantid
# Workbench Project Save Dialog

**Description of work.**
This PR adds a dialog when saving a project which allows the user to choose whether to save all workspaces or only the ones which have been altered, i.e. their history contains more than Load.

![image](https://user-images.githubusercontent.com/52415735/82657593-83f3d080-9c1d-11ea-80a0-2540c55ee40f.png)

**To test:**
Run this script: 
```
from mantid.simpleapi import *

unaltered_ws = Load('EMU00019489')
altered_ws = Rebin('unaltered_ws', 5)

plotSpectrum(['unaltered_ws'], indices=[0])
plotSpectrum(['altered_ws'], indices=[0])
plotSpectrum(['altered_ws', 'unaltered_ws'], indices=[1])
```
Go to File -> Save Project as... and you should now see the dialog where you can choose the file location and whether to save all workspaces or altered ones only.

Save the project with 'Save All Workspaces' selected. Close Workbench and reopen it. Do File -> Open Project and select the project you just saved. All the workspaces and plots should be present. Make some change to the project, such as loading another workspace or closing a plot, and go to File -> Save Project. Close Workbench and reopen it, open the project and it should have remembered your choice to save all workspaces.

Close Workbench and reopen it and rerun the script. This time do File -> Save Project as..., select 'Save Altered Workspaces Only'. Close Workbench and reopen it, and open this project. `unaltered_ws` and the plots of it should no longer be there.

If you load another workspace and do File -> Save Project, the dialog should not appear. Now when you close and reopen Workbench and open the project, the newly loaded workspace should not be there as it remembered your choice to only save the altered workspace.

Basically do a lot of messing around with saving and loading projects and make sure everything works as expected.

Fixes #27472 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
